### PR TITLE
Fix start_tile_id when work is split over batches in MatmulMultiCoreConfig

### LINKED ISSUE #34708: Transpose Matmul fusion causes training loss regression
### Component / Area

matmul

### Issue Type (optional)

Bad Outputs

### Observed

After merge of #32295, training does not converge on nanogpt training after 5000 steps.
After isolating every transpose matmul operation performed during nanogpt training, one of them fails i.e. the results differ from the composite version.

### Expected

All fused transpose matmul operations should yield the same result as composite transpose matmul.
For the test case:
```
a.shape = [64, 6, 256, 256]
b.shape = [64, 6, 256, 64]
transpose_a = true
transpose_b = false
```
The results are not exact.
Other shapes work correctly.

### 1. Steps (exact commands)

```
git checkout zbaczewski/tranpose-matmul-bug
./build_metal.sh -b Release --build-tt-train
./build/tt-train/tests/ttml_tests --gtest_filter="*MatMulManualVsFlagTest*"
```

### 2. Input data / link or description

None

### 3. Frequency

Always

### 1. Software Versions

Ubuntu 22.04

### 2. Hardware Details

Wormhole N150

### Is this a regression?

Yes

### Regression Details

_No response_

### Logs & Diagnostics

```
...
[ RUN      ] MatMulTransposeTests/MatMulManualVsFlagTest.ManualTransposeMatchesFlag/1
y_manual shape: { 64,   6, 256,  64}
y_flag shape: { 64,   6, 256,  64}
Number of differing elements: 6171166 / 6291456
/home/ubuntu/tt-metal/tt-train/tests/ttnn_fixed/matmuls_test.cpp:301: Failure
Expected equality of these values:
  y_manual
    Which is: { 4.28125, -0.55078125, -4.46875, 6.84375, 8.1875, -7.03125, 7.875, -5.15625, 5.1875, 2.15625, -1.265625, -5.96875, -7.90625, 5.9375, -3.65625, 1.4375, -7.375, 0.357421875, -2.75, 6.6875, 10.1875, -8.75, 2.75, -5.90625, 1.28125, 6.09375, -0.083984375, 1.6484375, -2.375, 2.859375, 2.015625, 3.078125, ... }
  y_flag
    Which is: { 4.28125, -0.55078125, -4.46875, 6.84375, 8.1875, -7.03125, 7.875, -5.15625, 5.1875, 2.15625, -1.265625, -5.96875, -7.90625, 5.9375, -3.65625, 1.4375, -7.375, 0.357421875, -2.75, 6.6875, 10.1875, -8.75, 2.75, -5.90625, 1.28125, 6.09375, -0.083984375, 1.6484375, -2.375, 2.859375, 2.015625, 3.078125, ... }
[  FAILED  ] MatMulTransposeTests/MatMulManualVsFlagTest.ManualTransposeMatchesFlag/1, where GetParam() = a=[64,6,256,256]_b=[64,6,256,64]_transA=1_transB=0 (4150 ms)
...
```

Should be 0 / 6291456

### Priority

None

### Impact

tt-train team is affected by this issue.

### LINKED ISSUE #34811: L1 Memory Overflow in linear_op Backward Pass with Certain Vocabulary Sizes
The backward pass of ttml.ops.linear.linear_op crashes with an L1 memory overflow error for specific vocabulary sizes when using the hardcoded core_grid{7, 8} in ttnn_fixed::matmul. The issue is non-obvious as some larger vocab sizes work while smaller ones fail.

```
import ttml
import numpy as np

seq_len = 800
emb_dim = 384

def test_emb_backward(vocab_size):
    assert emb_dim % 32 == 0
    assert vocab_size % 32 == 0

    hidden = np.random.randn(1, 1, seq_len, emb_dim).astype(np.float32)
    lm_head_weight = np.random.randn(1, 1, vocab_size, emb_dim).astype(np.float32)

    hidden_ttml = ttml.autograd.Tensor.from_numpy(hidden, ttml.Layout.TILE, ttml.autograd.DataType.BFLOAT16)
    lm_head_ttml = ttml.autograd.Tensor.from_numpy(lm_head_weight, ttml.Layout.TILE, 
                                                   ttml.autograd.DataType.BFLOAT16)

    logits = ttml.ops.linear.linear_op(hidden_ttml, lm_head_ttml, None)
    logits.backward(False)
    
# doesn't crash    
test_emb_backward(80000) 

# TT_THROW: Statically allocated circular buffers on core range [(x=0,y=0) - (x=6,y=7)] 
# grow to 1528096 B which is beyond max L1 size of 1499136 B (assert.hpp:103)
test_emb_backward(75968) 

# TT_THROW: Statically allocated circular buffers on core range [(x=0,y=0) - (x=6,y=7)] 
# grow to 1544480 B which is beyond max L1 size of 1499136 B (assert.hpp:103)
test_emb_backward(151936) 
```

Replacing ttnn::CoreGrid{7, 8} -> std::nullopt eliminates error, but reduce performance. I think the issue might be related to the assumption about the entire grid range when calculating the buffer size in matmul.

Forward pass is working, it crashing only on backward.

This affects training of Qwen3 models - Qwen3 uses vocab_size=151936, causing training failures while Llama (vocab_size=128256) works

### PR DESCRIPTION
**Pull request by nsorabaTT**
### Ticket
#34708
#34811

### Problem description
Consider a matmul of shapes [b,s,m,k] @ [b,s,m,n]. `MatmulMultiCoreReuseProgramConfig` splits the workload over the available cores in two possible ways:
1. Split `b*s` number of batches among all the available cores. Each core completes all parts of the batch.
2. Split `M` over all available cores. Each core completes a subset of output for each batch.

The tile offsets are different depending on whether the cores are split by (1) or (2). This is further exacerbated when the transpose flags are enabled, since the tile idx for the first tile read in by each core will be different.

### What's changed
1. Added a flag to detect the way work is split.
2. Added logic to correct the `start_tile_id` for in0 and in1 based on work split and transpose flags.
3. Fix estimated size for CBs to include the transpose_CB that is used when inputA needs to be transposed. 

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes https://github.com/tenstorrent/tt-metal/actions/runs/20469829488
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable) https://github.com/tenstorrent/tt-metal/actions/runs/20469832611
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable) https://github.com/tenstorrent/tt-metal/actions/runs/20466884236/job/58813346372
- [x] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable) https://github.com/tenstorrent/tt-metal/actions/runs/20472479189
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable)
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml) job passes
- [x] New/Existing tests provide coverage for changes
- [x] L2 Nightly https://github.com/tenstorrent/tt-metal/actions/runs/20454747627 (Failure matches main, unrelated)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_optimized_program_factory.cpp [hunk 1]
```diff
@@ -458,20 +456,41 @@ tt::tt_metal::operation::ProgramWithCallbacks create_program(
     }
     const auto cores = grid_to_cores(num_cores, core_range.x, core_range.y, row_major);
 
+    // There are two cases:
+    // (a) each core processes all M,N dimensions of b number of batches
+    // (b) each core processes a subset of M dimension of b number of batches
```
[the remark this answers is not in the document]
Comment (nsorabaTT) [reply ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_optimized_program_factory.cpp]: This config does not split across N dimension, so the comment is correct and copilot's suggestion is incorrect.

**Review by bbradelTT — APPROVED** [review]
